### PR TITLE
[Webgl2] Add ssd_mobilenet

### PR DESCRIPTION
**Pull request by ganwenyao**
Fix https://github.com/intel/webml-polyfill/issues/16

### CODE REVIEW

**Comment by huningxin** [comment]
Thanks! I will review the PR ASAP.

File: examples/ssd_mobilenet/model/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+The model and label files are:
+
+```txt
+coco_labels_list.txt
+ssd_mobilenet.tflite
+```
+
+Check out [TensorFlow Lite Models](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) for details.
+
+This modle is conver from [Tensorflow SSD MobileNet model](http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_2017_11_17.tar.gz). You can use the following commands to convert your own model.
```
Comment (huningxin) [inline examples/ssd_mobilenet/model/README.md]: "conver" -> "converted"

File: examples/ssd_mobilenet/MobileNet.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    this._compilation.setPreference(this._nn.PREFER_FAST_SINGLE_ANSWER);
+    await this._compilation.finish();
+    this._execution = await this._compilation.createExecution();
+  }
+
+  async compute(inputTensor, outputBoxTensor, outputClassScoresTensor) {
+    // this._execution._inputs = [];
+    // this._execution._outputs = [];
+    this._execution.setInput(0, inputTensor);
+    let outH = [1083, 600, 150, 54, 24, 6];
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: "let" -> "const"

File: examples/ssd_mobilenet/MobileNet.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    let boxLen = 4;
+    let classLen = 91;
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: ditto

Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: And please document where these constants come from, as you modified the net outputs.

File: examples/ssd_mobilenet/model/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+--input=${download_dir}/ssd_mobilenet_v1_coco_2017_11_17/frozen_inference_graph.pb \
+--output=${out_dir}/frozen_inference_graph_stripped.pb --frozen_graph=True \
+--input_names=Preprocessor/sub \
+--output_names=\
+"BoxPredictor_0/BoxEncodingPredictor/BiasAdd,BoxPredictor_0/ClassPredictor/BiasAdd,\
+BoxPredictor_1/BoxEncodingPredictor/BiasAdd,BoxPredictor_1/ClassPredictor/BiasAdd,\
+BoxPredictor_2/BoxEncodingPredictor/BiasAdd,BoxPredictor_2/ClassPredictor/BiasAdd,\
+BoxPredictor_3/BoxEncodingPredictor/BiasAdd,BoxPredictor_3/ClassPredictor/BiasAdd,\
+BoxPredictor_4/BoxEncodingPredictor/BiasAdd,BoxPredictor_4/ClassPredictor/BiasAdd,\
+BoxPredictor_5/BoxEncodingPredictor/BiasAdd,BoxPredictor_5/ClassPredictor/BiasAdd" \
```
Comment (huningxin) [inline examples/ssd_mobilenet/model/README.md]: Please document why you need change the output. For example, current WebML API doesn't support "squeeze" operation.

Comment (ganwenyao) [reply examples/ssd_mobilenet/model/README.md]: I add some notes.

File: examples/ssd_mobilenet/MobileNet.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+   * https://github.com/tensorflow/models/blob/master/research/object_detection/builders/model_builder.py
+   * https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_mobilenet_v1_coco.config
+   */
+  async compute(inputTensor, outputBoxTensor, outputClassScoresTensor) {
+   /**
+    * Object detection feature extractors usually are built by stacking two components:
+    * a base feature extractor such as mobilenet and a feature map generator.
+    * In our example, ssd_mobilenet use 2 feature maps('Conv2d_11_pointwise', 'Conv2d_13_pointwise') from mibilenet 
+    * and 4 feature maps generated from the base feature map('Conv2d_13_pointwise'). 
+    * The sizes of our 6 feature maps are [(19,19), (10,10), (5,5), (3.3), (2,2), (1,1)]
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: `(3.3)` -> `(3,3)`

File: examples/ssd_mobilenet/MobileNet.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+  /**
+   * See tensorflow ssd_mobilenet_v1 example for details:
+   * https://github.com/tensorflow/models/blob/master/research/object_detection/builders/model_builder.py
+   * https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_mobilenet_v1_coco.config
+   */
+  async compute(inputTensor, outputBoxTensor, outputClassScoresTensor) {
+   /**
+    * Object detection feature extractors usually are built by stacking two components:
+    * a base feature extractor such as mobilenet and a feature map generator.
+    * In our example, ssd_mobilenet use 2 feature maps('Conv2d_11_pointwise', 'Conv2d_13_pointwise') from mibilenet
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: 'mibilenet' -> 'mobilenet'

File: examples/ssd_mobilenet/MobileNet.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+   */
+  async compute(inputTensor, outputBoxTensor, outputClassScoresTensor) {
+   /**
+    * Object detection feature extractors usually are built by stacking two components:
+    * a base feature extractor such as mobilenet and a feature map generator.
+    * In our example, ssd_mobilenet use 2 feature maps('Conv2d_11_pointwise', 'Conv2d_13_pointwise') from mibilenet 
+    * and 4 feature maps generated from the base feature map('Conv2d_13_pointwise'). 
+    * The sizes of our 6 feature maps are [(19,19), (10,10), (5,5), (3.3), (2,2), (1,1)] 
+    * and each location of the feature map predicts 6 anchors, so the total number of output anchors is 
+    * 19^2*6 + 10^2*6 + 5^2*6 + 3^2*6 + 2^2*6 + 1^2*6 = 1083 + 600 + 150 + 54 + 24 + 6 = 1917.
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: According to config, the lowest layer only has 3 anchors, so `19^2*6 = 1083`.

File: examples/ssd_mobilenet/MobileNet.js (hunk 5 above)
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: `(3.3)` -> `(3,3)`

File: examples/ssd_mobilenet/MobileNet.js (hunk 7 above)
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: The lowest layer only has 3 anchors. So `19^2*3 = 1083`

File: examples/ssd_mobilenet/MobileNet.js (hunk 6 above)
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: typo 'mibilenet'

File: examples/ssd_mobilenet/utils.js [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+    let areaSum = area1 + area2 - intersection
+    if (areaSum === 0) {
+      throw new Error('[IOU] areaSum can not be 0!');
+    }
+    let IOU = intersection / areaSum;
+    return IOU;
+  }
+
+  /**
+  * Generate anchors
```
Comment (huningxin) [inline examples/ssd_mobilenet/utils.js]: Please add comment by referring to `create_ssd_anchors` of https://github.com/tensorflow/models/blob/master/research/object_detection/anchor_generators/multiple_grid_anchor_generator.py

Comment (huningxin) [reply examples/ssd_mobilenet/utils.js]: And please add referring to `ssd_anchor_generator` of https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_mobilenet_v1_coco.config

File: examples/ssd_mobilenet/utils.js [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+      if (interpolated_scale_aspect_ratio > 0.0) {
+        layer_box_specs.push([Math.sqrt(scale * scale_next),
+                                interpolated_scale_aspect_ratio])
+      }
+    }
+    box_specs_list.push(layer_box_specs);
+  });
+
+  let anchors = [];
+  let anchorCord4 = [];
```
Comment (huningxin) [inline examples/ssd_mobilenet/utils.js]: Remove unused `anchorCord4` and following commented out lines.

File: examples/ssd_mobilenet/MobileNet.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
+let nnPolyfill, nnNative;
+if (navigator.ml.isPolyfill) {
+  nnNative = null;
+  nnPolyfill = navigator.ml.getNeuralNetworkContext();
+} else {
+  nnNative = navigator.ml.getNeuralNetworkContext();
+  nnPolyfill = navigator.ml_polyfill.getNeuralNetworkContext();
+}
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: Please remove above code as they are in https://github.com/intel/webml-polyfill/blob/master/examples/util/base.js

File: examples/ssd_mobilenet/MobileNet.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+class MobileNet {
```
Comment (huningxin) [inline examples/ssd_mobilenet/MobileNet.js]: I would suggest to rename the class name and file name to "SsdMobileNet".

Comment (ganwenyao) [reply examples/ssd_mobilenet/MobileNet.js]: I updated my code.

File: examples/ssd_mobilenet/model/README.md (hunk 4 above)
Comment (huningxin) [reply examples/ssd_mobilenet/model/README.md]: looks good.

**Comment by huningxin** [comment]
Looks very good. Thanks for the PR!

**Pull request merged**